### PR TITLE
feat(scm): ignore scm data when scanning

### LINKED ISSUE #1251: Automatically ignore SCM data when scaning
### Description

When scanning an upload, I need to automatically ignore SCM specific data (Git, SVN, ... files).
Issue #703 hints at this, but also at ignoring other parts of the upload, so I would rather create a new issue concerning the SCM data specifically.

Not sure about the exact list of SCM solutions to be integrated (at least Git, SVN and TFS)

#### Steps followed and expected result

When uploading a new package for scanning (any type, really: from file, server, URL, SCM, TFS), a new option would enable the feature. See image below, where a new `4. Ignore SCM files` has been inserted.
When selected, the scan automatically discards all files related to known SCM solutions.

#### Screenshots

image

### PR DESCRIPTION
**Pull request by ChristopheRequillart**
### Description 
Implement ignore scm data when scanning an upload (see #1251)

### Changes
1. Add a new column scm in upload table
1. Add a new option to enable this feature when adding a new upload. 
1. Add a new function in ununpack/agent/utils.c to test SCM data file (see SCM_REGEX constant to search git, svn, TFS, mercurial).

## How to test
Add a new upload with scm data files. Enable this new option to not scanning SCM data files.

This is a duplicate of  #1264 to resolve multiple commits problem
Closes #1251

### CODE REVIEW

**Comment by GMishx** [comment]
Can you please change the commit message on your second commit (0d3af64ca5f0161a2dc7dd04a683b8d6b30b9d20) and make it similar to first one (as per the contribution guidelines)?

Use
1. `git commit --amend`
1. Change message
1. Do a force push by `git push --force-with-lease`

**Comment by ChristopheRequillart** [comment]
Git don’t know the option --force-with-lease
```
usage: git push [<options>] [<repository> [<refspec>...]]

    -v, --verbose         be more verbose
    -q, --quiet           be more quiet
    --repo <repository>   repository
    --all                 push all refs
    --mirror              mirror all refs
    --delete              delete refs
    --tags                push tags (can't be used with --all or --mirror)
    -n, --dry-run         dry run
    --porcelain           machine-readable output
    -f, --force           force updates
    --recurse-submodules[=<check>]
                          control recursive pushing of submodules
    --thin                use thin pack
    --receive-pack <receive-pack>
                          receive pack program
    --exec <receive-pack>
                          receive pack program
    -u, --set-upstream    set upstream for git pull/status
    --progress            force progress reporting
    --prune               prune locally removed refs
    --no-verify           bypass pre-push hook
    --follow-tags         push missing but relevant tags
```

**Comment by GMishx** [comment]
Maybe then a normal force will do (`-f` or `--force`)

**Comment by shaheemazmalmmd** [comment]
@ChristopheRequillart : Can you please use the existing "jq_cmd_args" from jobqueue table for adding information about SCM instead of creating a column in Upload table.

File: src/www/ui/template/include/upload.html.twig [hunk 1]
```diff
@@ -35,6 +35,9 @@
           <input type="text" name="{{ descriptionInputName }}" size="120" value="{{ descriptionInputValue }}"/>
         </label>
       </li>
+      li>
```
Comment (shaheemazmalmmd) [inline src/www/ui/template/include/upload.html.twig]: "<li>" please correct the tag

Comment (ChristopheRequillart) [reply src/www/ui/template/include/upload.html.twig]: OK corrected

**Comment by ChristopheRequillart** [comment]
> @ChristopheRequillart : Can you please use the existing "jq_cmd_args" from jobqueue table for adding information about SCM instead of creating a column in Upload table.

In ./src/www/ui/page/Upload*Page.php, I can easily initialize jq_cmd_args while calling JobQueueAdd().
How find which arguments are currently used?
But in ununpack tool, in DBInsertUploadTree() function, there is only one Upload_PK variable and a ContainerInfo. How do you find a single jq_cmd_args with this?  `SELECT jq_cmd_args FROM jobqueue LEFT OUTER JOIN job ON jobqueue.jq_job_fk = job.job_pk WHERE job.job_upload_fk = Upload_PK` can get it but not a single one.

**Comment by shaheemazmalmmd** [comment]
**In ./src/www/ui/page/Upload*Page.php, I can easily initialize jq_cmd_args while calling JobQueueAdd().
How find which arguments are currently used?**

* You can add the jq_cmd_args for ununpack agent. 

**But in ununpack tool, in DBInsertUploadTree() function, there is only one Upload_PK variable and a ContainerInfo. How do you find a single jq_cmd_args with this? `SELECT jq_cmd_args FROM jobqueue LEFT OUTER JOIN job ON jobqueue.jq_job_fk = job.job_pk WHERE job.job_upload_fk = Upload_PK` can get it but not a single one.**

* use this query to get the single argument from database. `SELECT jq_cmd_args  FROM jobqueue LEFT OUTER JOIN job ON jobqueue.jq_job_fk = job.job_pk WHERE jobqueue.jq_type = 'ununpack' AND job.job_upload_fk = 6 ORDER BY job_queued DESC LIMIT 1;`

**Comment by ChristopheRequillart** [comment]
>     * You can add the jq_cmd_args for ununpack agent.
Sure but I have to find a new argument to put in jq_cmd_args. This new argument needs to be unique and not be used yet. How to find the list of current arguments used?

>     * use this query to get the single argument from database. `SELECT jq_cmd_args  FROM jobqueue LEFT OUTER JOIN job ON jobqueue.jq_job_fk = job.job_pk WHERE jobqueue.jq_type = 'ununpack' AND job.job_upload_fk = 6 ORDER BY job_queued DESC LIMIT 1;`

Are you sure it’s the best option? The SCM information will be duplicate n times. In upload table this information is unique.

**Comment by shaheemazmalmmd** [comment]
**Sure but I have to find a new argument to put in jq_cmd_args. This new argument needs to be unique and not be used yet. How to find the list of current arguments used?**

* arguments are specific to agents here and ununpack doesn't use any argument from ui. so you can choose any.

**Are you sure it’s the best option? The SCM information will be duplicate n times. In upload table this information is unique.**

* yes i hope. if you find better option please add.

**Comment by ChristopheRequillart** [comment]
>     * yes i hope. if you find better option please add.

For me, better option is one new column in upload table. As already done.

**Comment by mcjaeger** [comment]
> > ```
> > * You can add the jq_cmd_args for ununpack agent.
> > ```
> 
> Sure but I have to find a new argument to put in jq_cmd_args. This new argument needs to be unique and not be used yet. How to find the list of current arguments used?

why new argument? it is just natural to put this into the unpack agent command line arguments. That appears to be missing somehow, but at least that could be done in the job table?


> > ```
> > * use this query to get the single argument from database. `SELECT jq_cmd_args  FROM jobqueue LEFT OUTER JOIN job ON jobqueue.jq_job_fk = job.job_pk WHERE jobqueue.jq_type = 'ununpack' AND job.job_upload_fk = 6 ORDER BY job_queued DESC LIMIT 1;`
> > ```
> 
> Are you sure it’s the best option? The SCM information will be duplicate n times. In upload table this information is unique.

AFAIK, the information would be only stored for the unpack agent - where would you see it `n` times

**Comment by ChristopheRequillart** [comment]
> > > ```
> > > * You can add the jq_cmd_args for ununpack agent.
> > > ```
> > 
> > 
> > Sure but I have to find a new argument to put in jq_cmd_args. This new argument needs to be unique and not be used yet. How to find the list of current arguments used?
> 
> why new argument? it is just natural to put this into the unpack agent command line arguments. That appears to be missing somehow, but at least that could be done in the job table?
> 
I don’t understand. I need to inform unpack agent. So I have to put something it jq_cmd_args. It’s what I call an argument.

> > > ```
> > > * use this query to get the single argument from database. `SELECT jq_cmd_args  FROM jobqueue LEFT OUTER JOIN job ON jobqueue.jq_job_fk = job.job_pk WHERE jobqueue.jq_type = 'ununpack' AND job.job_upload_fk = 6 ORDER BY job_queued DESC LIMIT 1;`
> > > ```
> > 
> > 
> > Are you sure it’s the best option? The SCM information will be duplicate n times. In upload table this information is unique.
> 
> AFAIK, the information would be only stored for the unpack agent - where would you see it `n` times

How many lines have you in jobqueue table?

**Comment by ChristopheRequillart** [comment]
> > _Sure but I have to find a new argument to put in jq_cmd_args. This new argument needs to be unique and not be used yet. How to find the list of current arguments used?_*

@shaheemazmalmmd 
In ./src/www/ui/page/UploadFilePage.php, there isn't call to JobQueueAdd(). How to do it?

**Comment by GMishx** [comment]
> In ./src/www/ui/page/UploadFilePage.php, there isn't call to JobQueueAdd(). How to do it?

You can find the [`postUploadAddJobs`](https://fossology.github.io/d0/d59/UploadPageBase_8php_source.html#l00107) function which performs this job and [adds ununpack as dependency](https://fossology.github.io/d0/d59/UploadPageBase_8php_source.html#l00122). You can pass the argument required by ununpack agent as a `jqarg` accepted by [`doAgentAdd`](https://fossology.github.io/de/d33/classFossology_1_1Lib_1_1Plugin_1_1AgentPlugin.html#a7ec0c93c1eb3f05e0fb65da35334c6d9) function (which inturn calls [`JobQueueAdd`](https://fossology.github.io/df/d55/common-job_8php.html#abb0d47238f8917874543043be4b0542a)).

**Comment by ChristopheRequillart** [comment]
> > In ./src/www/ui/page/UploadFilePage.php, there isn't call to JobQueueAdd(). How to do it?
> 
> You can find the [`postUploadAddJobs`](https://fossology.github.io/d0/d59/UploadPageBase_8php_source.html#l00107) function which performs this job and [adds ununpack as dependency](https://fossology.github.io/d0/d59/UploadPageBase_8php_source.html#l00122). You can pass the argument required by ununpack agent as a `jqarg` accepted by [`doAgentAdd`](https://fossology.github.io/de/d33/classFossology_1_1Lib_1_1Plugin_1_1AgentPlugin.html#a7ec0c93c1eb3f05e0fb65da35334c6d9) function (which inturn calls [`JobQueueAdd`](https://fossology.github.io/df/d55/common-job_8php.html#abb0d47238f8917874543043be4b0542a)).

Are you sure it is in jqargs? Because as you said ununpack is a dependency of adj2nest. So it is $dependency when implicitAgentAdd() is called in doAgentAdd(), isn’t it?

**Comment by shaheemazmalmmd** [comment]
> > _Sure but I have to find a new argument to put in jq_cmd_args. This new argument needs to be unique and not be used yet. How to find the list of current arguments used?_*

>> >>_@shaheemazmalmmd 
In ./src/www/ui/page/UploadFilePage.php, there isn't call to JobQueueAdd(). How to do it?

You can directly get the value of SCM here https://github.com/fossology/fossology/blob/master/src/ununpack/ui/agent-unpack.php using "GetParm" or "$request". 

Now can pass the argument through this function doAgentAdd($jobId, $uploadId, &$errorMsg, $dependencies, $jqargs = "", $jq_cmd_args = null);

**Pull request merged**